### PR TITLE
Issue #570: Move logic of unpersistedbytes to bufferedchannel

### PR DESCRIPTION
**Pull request by reddycharan**
Descriptions of the changes in this PR:

This is < sub-task2 > of Issue #570.

https://github.com/apache/bookkeeper/commit/26b09abb4202362ca37d6944ce75eb2a3309dc3c
introduced the flushEntrylogBytes factor. But it is structurally correct to have this
logic in BufferedChannel, rather than in EntryLogger, since it is paramter of BufferedChannel.

Master Issue: #570

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java [hunk 1]
```diff
@@ -110,11 +150,15 @@ public long getFileChannelPosition() {
      * @throws IOException if the write or sync operation fails.
      */
     public void flush(boolean shouldForceWrite) throws IOException {
+        flush(shouldForceWrite, false);
+    }
+
+    public void flush(boolean shouldForceWrite, boolean forceMetadata) throws IOException {
         synchronized (this) {
```
[the remark this answers is not in the document]
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: Looking at flushInternal(), that method should be synchronized. I would synchronize flushInternal, and rename it to flush(), then kill these two methods completely. Where they are being called, if the user wants to force, they should call forceWrite directly (should probably be renamed force too).

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java [hunk 2]
```diff
@@ -138,6 +182,20 @@ public long forceWrite(boolean forceMetadata) throws IOException {
...
+         * unpersistedBytes to the current number of bytes in writeBuffer.
+         *
+         * since we are calling fileChannel.force, bytes which are written to
+         * filechannel (system filecache) will be persisted to the disk. So we
+         * dont need to consider those bytes for setting value to
+         * unpersistedBytes.
+         *
+         */
+        synchronized (this) {
+            unpersistedBytes.set(writeBuffer.readableBytes());
```
Comment (ivankelly) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: unpersistedBytes is called before the force, so the unpersisted bytes here is actually just the unflushed bytes. If this is correct, then unpersistedBytes should be updated in flushInternal. It can even be done by decrementAndGet

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java [hunk 3]
```diff
@@ -70,28 +96,42 @@ public void close() throws IOException {
...
+                copied += bytesToCopy;
+
+                // if we have run out of buffer space, we should flush to the
+                // file
+                if (!writeBuffer.isWritable()) {
+                    flushInternal();
+                }
+            }
+            position.addAndGet(copied);
+            unpersistedBytes.addAndGet(copied);
```
Comment (merlimat) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: Why do we need `AtomicInteger` if this is incremented withing a synchronized block?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerManagerFactory.java [hunk 4]
```diff
@@ -94,7 +94,7 @@ LedgerUnderreplicationManager newLedgerUnderreplicationManager()
      * @param lm
      *            Layout manager
      */
```
[the remark this answers is not in the document]
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerManagerFactory.java]: yes it is unrelated, but it is giving me compile time error in eclipse. So its needs to be fixed. Someone needs to be fixed. I'm trying to fix it in my other PR. If that PR goes first then this is not needed.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java (hunk 3 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: I'm exposing this variable's value through getUnpersistedBytes() method for testing purpose. Also I feel it is more safer to use atomic variables when it is accessed/modified by multiple threads.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java [hunk 5]
```diff
@@ -188,4 +246,12 @@ public synchronized void clear() {
         super.clear();
         writeBuffer.clear();
     }
-}
+
+    public synchronized int getNumOfBytesInWriteBuffer() {
```
[the remark this answers is not in the document]
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: actually I'm not seeing that method being used by either product code or by testcode, so will consider removing it (not sure why I added, probably used it in some testcode earlier). 

But if this is needed, then it should be synchronized because it is returning/getting the state of shared resource - "writeBuffer". "writeBuffer" is accessed within synchronized block in all other places in this class.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: yeah, it makes sense to add explicit 'synchronized' to flushInternal method, though it is currently called by the callers which are in synchronized block, so that it wont be misused by future callers.

Other than that, I want to make minimal changes to the code while adding the functionality I intend to add. I don't want to do code refactoring if it is not related to this change / caused by this change..

[the remark this answers is not in the document]
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: +1 for minimal changes. since flushInternal is private, it won't be used outside of this class.

[the remark this answers is not in the document]
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: I'm strongly -1 on adding another boolean parameter to this method. You're modifying now, so now is the time to do it cleanly.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: hmm, yes, it seems readableBytes() isn't thread-safe(), which is surprising.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: @ivankelly 

if you are breaking this method, you need to change all the places that use this method, which is effectively a "code refactor". what is the real point of doing this unrelated code refactor here, which sounds unrelated or not needed for the functionality here.

what is the real problem here about this method for you to give -1? a real performance issue, race condition? or just a coding style?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java (hunk 2 above)
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: in flushInternal call, "fileChannel.write(toWrite);" method is called. It just writes to the file buffer but it is not guaranteed to be persisted to the storage device, hence flushInternal shouldn't affect unpersistedBytes count in anyway. But where as in the forceWrite method we are calling fileChannel.force method which "forces any updates to this channel's file to be written to the storage device that contains it", hence here we should decrement the count of the bytes which are written to filechannel.

in summary, way i see this is like following : unpersistedbytes = writeBuffer.readableBytes + bytes written to filchannel (which are not persisted yet by calling filechannel.force)

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: boolean parameters are code smell. Code review isn't just about performance or race conditions.

[the remark this answers is not in the document]
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: well, I don't think it is code smell. It is okay to me, because the flags define verify clear behavior: flush the buffer and if forceWrite enable, force the channel with 'forceMetadata' flag. The method follows the java standard: https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileChannel.html#force(boolean)

so going back to your comment here:

if your -1 is about the flag, `forceMetadata` flag is following the standard : https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileChannel.html#force(boolean)

if your -1 is about to ask Charan to break the methods into two methods, as Charan explained and I pointed out, this breakdown is a "code refactor" which will touch all the methods using `flush`. It is unrelated to the change, which can be done in a separate PR, so it doesn't sound to be a blocker to the change here.

so which part are you really giving -1 to?

[the remark this answers is not in the document]
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: flushInternal just flushes writeBuffer to fileChannel (doesn't persists and moreover it is private method). forceWrite just forces filechannel (persists data written to filechannel so far, doesn't touches writeBuffer), flush calls flushInternal and provides option to whether do forceWrite or not.

So each method has different purpose. replacing flushinternal with flush is not correct thing to do from its expectations. And if user wants to do both flushing writeBuffer and forcewrites the filechannel, then calling just forceWrite is not right thing to do.

But again I don't see point in complicating things by changing the signatures of this commonly used methods and touch all the callers of them. It is extraneous.

[the remark this answers is not in the document]
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: > well, I don't think it is code smell.

It's code smell. Google "boolean parameter". You'll get a selection of hits explaining why they are bad, including a blog post from Martin Fowler and another from Robert C Martin who wrote the book on clean code.

> because the flags define verify clear behavior

Do they? In 6 months are you going to read flush(false, true) and know exactly what the flags are doing?

> The method follows the java standard:

A 16 year old API. It can hardly be held up as a gold standard. Still though, it's not forceMetadata in itself I'm strongly opposed to. Its adding *another* boolean flag, to a method that already has one, that I'm opposed to.

> if your -1 is about to ask Charan to break the methods into two methods, as Charan explained and I pointed out, this breakdown is a "code refactor" which will touch all the methods using flush.

There's no refactor involved here. ```flush(boolean,boolean)``` isn't even called, so if it's removed, you will only have update ```flush(boolean)```.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java (hunk 2 above)
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: ok, maybe add a comment that unpersistedBytes may overreport, but that's ok.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerManagerFactory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerManagerFactory.java]: What error does it give in eclipse?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: anyway, I would defer this to Charan. His code change works for me. I have also expressed my opinion here.

[the remark this answers is not in the document]
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: > There's no refactor involved here. flush(boolean,boolean) isn't even called, so if it's removed, you will only have update flush(boolean).

It is not true. I'm planning to use in my next subtask. You can check how this method is used in my comprehensive PR - https://github.com/apache/bookkeeper/pull/1201 . We had quite big conversation regarding why 'forceMetadata' flag is required - (final conclusion of the conversation) - http://mail-archives.apache.org/mod_mbox/bookkeeper-dev/201705.mbox/%3CCAO2yDyahWu0ezJmQ66sdOvZBUofQu%3D8CabPDeH5JTRzkMULUCg%40mail.gmail.com%3E This conversation happened in May2017 for full thread - http://mail-archives.apache.org/mod_mbox/bookkeeper-dev/201705.mbox/browser

[the remark this answers is not in the document]
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: refactor refers to existing code. The next subtask is not existing code, so calls in it to ```flush(boolean, boolean)```, can be replaced with calls to ```flush(); forceWrite(boolean);``` or ```flushAndForceWrite(boolean)```

[the remark this answers is not in the document]
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: I was not referring to next subtask code for code refactoring. I was trying to explain -"flush(boolean,boolean) isn't even called", that it will be used by my future task and it cann't be removed.

It is not reasonable to expect people who touch the code intending to add new functionality to fix all the shortcomings of the existing code. It is risky and burdensome to touch critical path (change signatures) for mere arguable code cleanliness.

[the remark this answers is not in the document]
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java]: >  fix all the shortcomings of the existing code.

You are adding a new method. I'm asking you not to add to issues that already exist.

> It is risky and burdensome to touch critical path (change signatures)

You don't have to change any signatures if you don't want to. Just call flush(false), and then forceWrite(true). flush() was just a suggestion as the false is redundant there. BufferedChannel should have a simple flush() operation that just flushing the buffer to the channel, which is exactly what flush(false) does. You don't have to add that though.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerManagerFactory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (reddycharan) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerManagerFactory.java]: Redundant 'final' modifier.	LedgerManagerFactory.java	/bookkeeper-server/src/main/java/org/apache/bookkeeper/meta	line 97	Checkstyle Problem

[the remark this answers is not in the document]
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/LedgerManagerFactory.java]: Ah, weird. I wonder why checkstyle doesn't pick it up as part of the build. Anyhow, really shouldn't be in this patch, but will let it pass. It's the other issues I'm more concerned about.

**Comment by sijie** [comment]
@ivankelly : I think @reddycharan already addressed your comment. Can you take a look? if there is no other comments, let's move forward with it.

**Comment by ivankelly** [comment]
lgtm +1. Thanks Charan. Going to merge.
